File: sample/src/main/java/vn/tiki/sample/di/AppModule.java
package vn.tiki.sample.di;

import android.arch.persistence.room.Room;
import android.content.Context;
import dagger.Module;
import dagger.Provides;
import java.io.File;
import javax.inject.Singleton;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;
import vn.tiki.sample.BuildConfig;
import vn.tiki.sample.api.ApiService;
import vn.tiki.sample.database.AppDatabase;
import vn.tiki.sample.database.CartDao;
import vn.tiki.sample.model.UserModel;

/**
 * Created by Giang Nguyen on 8/25/17.
 */
@Module
public class AppModule {

  private final Context appContext;

  public AppModule(Context appContext) {
    this.appContext = appContext;
  }

  @Singleton
  @Provides
  public UserModel provideUserModel() {
    return new UserModel();
  }

  @Singleton
  @Provides
  ApiService provideApiService(OkHttpClient okHttpClient) {
    return new Retrofit.Builder()
        .baseUrl(BuildConfig.BASE_URL)
        .client(okHttpClient)
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .build()
        .create(ApiService.class);
  }

  @Singleton
  @Provides
  AppDatabase provideAppDatabase() {
    return Room.databaseBuilder(appContext, AppDatabase.class, "shopping").build();
  }

  @Provides
  File provideCacheDir() {
    return appContext.getCacheDir();
  }

  @Provides
  CartDao provideCartDao(AppDatabase appDatabase) {
    return appDatabase.cartDao();
  }

  @Singleton
  @Provides
  OkHttpClient provideOkHttpClient() {
    HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
    logging.setLevel(HttpLoggingInterceptor.Level.BODY);
    return new OkHttpClient.Builder()
        .addInterceptor(logging)
        .build();
  }
}

File: sample/src/main/java/vn/tiki/sample/di/AppModule.kt
package vn.tiki.sample.di

import android.arch.persistence.room.Room
import android.content.Context
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import dagger.Module
import dagger.Provides
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import vn.tiki.sample.BuildConfig
import vn.tiki.sample.api.ApiService
import vn.tiki.sample.database.AppDatabase
import vn.tiki.sample.database.CartDao
import vn.tiki.sample.entity.MyAdapterFactory
import vn.tiki.sample.model.UserModel
import java.io.File
import javax.inject.Singleton

@Module
open class AppModule(private val appContext: Context) {

  @Singleton
  @Provides
  open fun provideUserModel(): UserModel {
    return UserModel()
  }

  @Singleton
  @Provides
  internal fun provideApiService(okHttpClient: OkHttpClient): ApiService {
    return Retrofit.Builder()
        .baseUrl(BuildConfig.BASE_URL)
        .client(okHttpClient)
        .addConverterFactory(GsonConverterFactory.create())
        .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
        .build()
        .create(ApiService::class.java)
  }

  @Singleton
  @Provides
  internal fun provideAppDatabase(): AppDatabase {
    return Room.databaseBuilder(appContext, AppDatabase::class.java, "shopping").build()
  }

  @Provides
  internal fun provideCacheDir(): File {
    return appContext.cacheDir
  }

  @Provides
  internal fun provideCartDao(appDatabase: AppDatabase): CartDao {
    return appDatabase.cartDao()
  }

  @Provides
  internal fun provideGson(): Gson {
    return GsonBuilder()
        .registerTypeAdapterFactory(MyAdapterFactory.create())
        .create()
  }

  @Singleton
  @Provides
  internal fun provideOkHttpClient(): OkHttpClient {
    val logging = HttpLoggingInterceptor()
    logging.level = HttpLoggingInterceptor.Level.BODY
    return OkHttpClient.Builder()
        .addInterceptor(logging)
        .build()
  }

}
